### PR TITLE
Design doc for "Scale Mode"

### PR DESCRIPTION
**Pull request by 2uasimojo**
Propose a design to teach hive to be able to scale to O(100k) spoke
clusters.

[HIVE-1862](https://issues.redhat.com//browse/HIVE-1862)

### CODE REVIEW

**Comment by 2uasimojo** [comment]
@abutcher @jnpacker @xiangjingli PTAL, thx!

File: docs/enhancements/scale-mode.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+Operations involving workloads will be performed in the control plane, using the client based on the controller manager, as normal.
+Other operations will be performed in the data plane, using a client constructed from the aforementioned kubeconfig.
+
+## HiveConfig
+HiveConfig will grow one additional field, `Spec.ScaleMode`, a boolean.
+When false, hive will operate in normal mode, managing and deploying to the (single) namespace indicated by `Spec.TargetNamespace` (which defaults to `hive` if unspecified).
+When true, hive will ignore `Spec.TargetNamespace` and operate as further described below.
+
+The existing `Ready` condition in the `Status.Conditions` list will continue to be used to indicate the health of the deployments.
+If *any* data plane is unhealthy, the condition's `Status` will be set to `False` and its `Message` will enumerate the unhealthy namespaces.
```
Comment (2uasimojo) [inline docs/enhancements/scale-mode.md:66]: As I started looking into the code around this, I realized it wasn't going to work very well. Rewriting and adding this option to Alternatives (with others).

Comment (2uasimojo) [reply docs/enhancements/scale-mode.md]: ✓

File: docs/enhancements/scale-mode.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+As noted above, the consumer can ask hive to tear down the controllers for a data plane by removing the `data-plane` label from that namespace.
+Hive will clean up all and only those objects it created when deploying.
+(Notably, this does not include the `data-plane-kubeconfig` secret.)
+Hive will not destroy spoke clusters managed by the deployment in this namespace.
+
+Alternatively, the consumer can simply delete the namespace, which should cause all objects therein to be deleted and garbage collected as normal.
+
+#### ClusterDeployment-Related Workloads
+The **consumer** is responsible for managing (creating and destroying) namespaces in the control plane cluster parallel to (i.e. with the same name as) those in the data planes.
```
Comment (2uasimojo) [inline docs/enhancements/scale-mode.md:154]: ClusterPools will attempt to create namespaces for pool clusters. Perhaps this should sound more like, "Whoever creates the namespace for the ClusterDeployment must create it in both planes." I.e. the consumer should do it for a regular non-pool CD, but the clusterpool controller should do it for pool CDs.

File: docs/enhancements/scale-mode.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+- Stand up the data plane.
+  (Note that hive doesn't care where the data plane actually lives.
+  In particular, it may or may not be a service running in this namespace.)
+- Create the `data-plane-kubeconfig` Secret in the namespace.
+- Label the namespace with `hive.openshift.io/data-plane=true`
+
+What with eventual consistency, it's not crucial that the above be done in order; but adding the label last will result in the smoothest transition.
+Otherwise HiveConfig may temporarily report `Failure` status for the namespace e.g. if the Secret doesn't exist yet.
+
+As noted above, the consumer can ask hive to tear down the controllers for a data plane by removing the `data-plane` label from that namespace.
```
Comment (abutcher) [inline docs/enhancements/scale-mode.md:148]: Are data planes removed from `Status` when they are deleted/de-labeled?

Comment (abutcher) [reply docs/enhancements/scale-mode.md:148]: They probably are as a function of the tear down.

File: docs/enhancements/scale-mode.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+## Overview
+hive-operator will learn how to deploy the controller stack (hive-controllers, hiveadmission, hive-clustersync) to more than one target namespace.
+In each such namespace it will expect to find a secret containing a kubeconfig providing access to a data plane.
+(The data plane must be globally unique to preclude separate controller stacks colliding attempting to reconcile the same object.)
+For each k8s CRUD operation hive needs to perform, the controller stack will decide whether the operation involves a workload (container) or not.
+Operations involving workloads will be performed in the control plane, using the client based on the controller manager, as normal.
+Other operations will be performed in the data plane, using a client constructed from the aforementioned kubeconfig.
+
+## HiveConfig
+HiveConfig.Spec will grow one additional field, `Spec.ScaleMode`, a boolean.
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply docs/enhancements/scale-mode.md:65]: Mm, that's a good idea.

File: docs/enhancements/scale-mode.md (hunk 3 above)
Comment (2uasimojo) [reply docs/enhancements/scale-mode.md:148]: Yes they are. The only way you would see them in hiveconfig status after they're unlabeled is if something goes wrong as we try to clean them up.

File: docs/enhancements/scale-mode.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (2uasimojo) [reply docs/enhancements/scale-mode.md]: Done.

File: docs/enhancements/scale-mode.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+hive-operator will learn how to deploy the controller stack (hive-controllers, hiveadmission, hive-clustersync) to more than one target namespace.
+In each such namespace it will expect to find a secret containing a kubeconfig providing access to a data plane.
+(The data plane must be globally unique to preclude separate controller stacks colliding attempting to reconcile the same object.)
+For each k8s CRUD operation hive needs to perform, the controller stack will decide whether the operation involves a workload (container) or not.
+Operations involving workloads will be performed in the control plane, using the client based on the controller manager, as normal.
+Other operations will be performed in the data plane, using a client constructed from the aforementioned kubeconfig.
+
+## HiveConfig
+HiveConfig.Spec will grow an additional substruct field, `Spec.ScaleMode`, to hold configuration for this new mode.
+Initially the a substruct will contain a single boolean field, `Enabled`.
```
Comment (abutcher) [inline docs/enhancements/scale-mode.md:66]: `s/a //` or replace with `ScaleMode`

```suggestion
Initially the `ScaleMode` substruct will contain a single boolean field, `Enabled`.
```

**Review by abutcher — COMMENTED** [review]
Looks good to me! One nit from the substruct bump

**Pull request merged**